test/utils.test.js

```diff
@@ -14,37 +38,367 @@ describe('utils', function() {
+      const jml = ['div', 'hello world'];
+      assert.equal(utils.getAttribute(jml, 'a'), undefined);
+    });
+
+    it('should throw an error when passed an invalid element', function() {
+      [5, false, undefined, null, {}].forEach(val => {
+        assert.throws(() => utils.getAttribute(val), SyntaxError);
+      });
+      assert.doesNotThrow(() => utils.getAttribute(''));
+      assert.doesNotThrow(() => utils.getAttribute(['']));
+    });
+  });
+
+  describe('#setAttribute', function() {
+    it('should export a function', function() {
+      assert.equal(typeof utils.setAttribute, 'function');
+    });
+
+    it('should set an attriute on an element without any', function() {
+      const jml = ['p', 'hello world'];
+      utils.setAttribute(jml, 'a', 1);
+      assert.deepEqual(jml, ['p', { a: 1 }, 'hello world']);
+    });
+
+    it('should set an attribute on an element with existing attributes', function() {
+      const jml = ['p', { a: 1 }, 'hello world'];
+      utils.setAttribute(jml, 'b', 2);
+      assert.deepEqual(jml, ['p', { a: 1, b: 2 }, 'hello world']);
+    });
+
+    it('should override an existing attribute value', function() {
+      const jml = ['p', { a: 1 }, 'hello world'];
+      utils.setAttribute(jml, 'a', 123);
+      assert.deepEqual(jml, ['p', { a: 123 }, 'hello world']);
+    });
+
+    it('should throw an error if element is a string', function() {
+      assert.throws(() => utils.setAttribute('hello'), Error);
+    });
+
+    it('should throw an error when passed an invalid element', function() {
+      [5, false, undefined, null, {}].forEach(val => {
+        assert.throws(() => utils.setAttribute(val), SyntaxError);
+      });
+      assert.doesNotThrow(() => utils.setAttribute(['']));
+    });
+  });
+
+  describe('#appendChild', function() {
+    it('should export a function', function() {
+      assert.equal(typeof utils.appendChild, 'function');
+    });
+
+    it('should throw an error if parent param cannot contain children', function() {
+      const parent = 'hello';
+      const child = 'world';
+      assert.throws(() => utils.appendChild(parent, child), SyntaxError);
+      assert.doesNotThrow(() => utils.appendChild(['p'], child));
+    });
+
+    describe('when appending a fragment', function() {
+      it('should append all elements', function() {
+        const jml = ['div'];
+        const fragment = ['',
+          ['em', 'hello'],
+          ['strong', 'world'],
+        ];
+        utils.appendChild(jml, fragment);
+        assert.deepEqual(jml, ['div', ['em', 'hello'], ['strong', 'world']]);
+      });
+
+      // @TODO - is this an intentional mutation?
+      it('should remove fragment contents', function() {
+        const jml = ['div'];
+        const fragment = ['',
+          ['em', 'hello'],
+          ['strong', 'world'],
+        ];
+        utils.appendChild(jml, fragment);
+        assert.deepEqual(fragment, []);
+      });
```

Comment:
I added this test when I realized the `fragment` was being muted (by calling [`child.shift()`](https://github.com/benjycui/jsonml.js/blob/master/lib/utils.js#L92)). I wasn't sure if this was intentional. Personally, I would argue that functions should be pure whenever possible and that the child param _should not_ be mutated.

Reply:
Here's what a fix would look like:
https://github.com/pgoldrbx/jsonml.js/compare/test/utils...pgoldrbx:fix/child-fragment-mutation?expand=1

Reply:
Yep, we had better to make function pure, and you can fix it in next PR, thanks.

Reply:
Filed: #7 
Patch: #8